File: leetcode/0652-find-duplicate-subtrees/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        serialize(root);
        return result;
    }

private:
    unordered_map<string, int> counts;
    vector<TreeNode*> result;

    string serialize(TreeNode* node) {
        if (node == nullptr) {
            return "#";
        }

        string left = serialize(node->left);
        string right = serialize(node->right);
        string key = to_string(node->val) + "," + left + "," + right;

        int count = ++counts[key];
        if (count == 2) {
            result.push_back(node);
        }

        return key;
    }
};

File: leetcode/0652-find-duplicate-subtrees/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private final Map<String, Integer> counts = new HashMap<>();
    private final List<TreeNode> result = new ArrayList<>();

    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        serialize(root);
        return result;
    }

    private String serialize(TreeNode node) {
        if (node == null) {
            return "#";
        }

        String left = serialize(node.left);
        String right = serialize(node.right);
        String key = node.val + "," + left + "," + right;

        int count = counts.merge(key, 1, Integer::sum);
        if (count == 2) {
            result.add(node);
        }

        return key;
    }
}
